neps/nep-0364.md

```diff
@@ -0,0 +1,203 @@
+fast machine code. Even when it may not be as fast as LLVM native produced code it could still be acceptable for
+these types of use cases (CPU intensive functions) and will certainly avoid the need of adding host functions.
+The effort of adding such a feature will be significantly higher than adding these host functions one by one.
+But on the other side, it will decrease the need of including more host functions in the future.
+
+Another alternative is to deal with the high cost of computing/verifying these signatures in some other manner.
+Decreasing the overall cost of gas and increasing the limits of gas available to attach to the contract could be a possibility.
+Introducing such modification for some contracts, and not for some others can be rather arbitrary
+and not straightforward in the implementation, but an alternative nevertheless.
+
+## Specification
+
+This NEP aims to introduce the following host function:
+
+```rust
+extern "C"{
+
+/// Verify an ED25519 signature given a message and a public key.
+/// Ed25519 is a public-key signature system with several attractive features
+///
+/// Proof of Stake Validator sets can contain different signature schemes.
+/// Ed25519 is one of the most used ones across blockchains, and hence it's importance to be added.
+/// For further reference, visit: https://ed25519.cr.yp.to
+/// # Returns
+/// - 1 if the signature was properly verified
+/// - 0 if the signature failed to be verified
+///
+/// # Cost
+///
+/// Each input can either be in memory or in a register. Set the length of the input to `u64::MAX`
+/// to declare that the input is a register number and not a pointer.
+/// Each input has a gas cost input_cost(num_bytes) that depends on whether it is from memory
+/// or from a register. It is either read_memory_base + num_bytes * read_memory_byte in the
+/// former case or read_register_base + num_bytes * read_register_byte in the latter. This function
+/// is labeled as `input_cost` below.
+///
+/// `input_cost(num_bytes_signature) + input_cost(num_bytes_message) + input_cost(num_bytes_public_key)
+///  ed25519_verify_base + ed25519_verify_byte * num_bytes_message`
+///
+/// # Errors
+///
+/// If the signature size is not equal to 64 bytes, or public key length is not equal to 32 bytes, contract execution is terminated with an error. 
+  fn ed25519_verify(
+    sig_len: u64,
+    sig_ptr: u64,
+    msg_len: u64,
+    msg_ptr: u64,
+    pub_key_len: u64,
+    pub_key_ptr: u64,
+  ) -> u64;
+```
+
+And a `rust-sdk` possible implementation could look like this:
+
+```rs
+
+pub fn ed25519_verify(sig: &ed25519::Signature, msg: &[u8], pub_key: &ed25519::Public) -> bool;
+
+```
+The current implementation is imported from the crate `ed25519-dalek`, version 1. It uses no feature flags
+except from the default one. This is the exact same setup used in the crate `near-crypto`.
```

Comment:
This sentence now seems redundant, given that the expected behavior is described below more thoroughly.